llvm/lib/Transforms/Coroutines/CoroSplit.cpp

```diff
@@ -1056,6 +1083,25 @@ void CoroCloner::create() {
   // Set up the new entry block.
   replaceEntryBlock();
 
+  // Turn symmetric transfers into musttail calls.
+  for (CallInst *ResumeCall : Shape.SymmetricTransfers) {
+    ResumeCall = cast<CallInst>(VMap[ResumeCall]);
+    ResumeCall->setCallingConv(NewF->getCallingConv());
+    if (TTI.supportsTailCallFor(ResumeCall)) {
+      // FIXME: Could we support symmetric transfer effectively without
+      // musttail?
+      ResumeCall->setTailCallKind(CallInst::TCK_MustTail);
+    }
+
+    // Put a 'ret void' after the call, and split any remaining instructions to
```

Comment:
if there are remaining instructions, this will silently elide them. I realize that there shouldn't be remaining instructions, but we could detect bugs like the PGO instrumentation + coro one (PR #71263) if that was asserted or WARNed or something more visible.

Reply:
I'm hesitant to add such a check. I don't think it should be illegal to have instructions after the llvm.coro.await.suspend.handle call, the semantics is just that those instructions won't execute -- similarly to how instructions after other function calls may not run. For example one could imagine some kind of coverage instrumentation or debug intrinsics or whatnot; those would not run, meaning that code path wasn't taken.

For the PGO issue, if I understand correctly the issue was that it prevented the musttail lowering, which won't happen anymore.

Reply:
It's not about legality, it's about the fact they wouldn't be executed being a symptom of a bug - e.g. earlier pass (like coverage) believing there is something to cover while there isn't. We could at minimum have some "under build with asserts" behavior that (except for trivial cases like debug/lifetime instructions) says something.

Reply:
I want the semantics to be well specified. Either the doc for this intrinsic should say that instructions between it and `llvm.coro.suspend` are not allowed (possibly with a list of exceptions) and we enforce it with an assert or the verifier; or it allows such instructions but says that they become unreachable as in the current version.

I'm in favor of the latter because it makes the IR more regular. We generally don't impose restrictions about what instructions are allowed between intrinsic calls. Instructions after a call not getting executed is not new, so I think what the current version of the patch does fits in well with the IR.

Reply:
Sorry for insisting on this, it's maybe because I got "bitten" before (with the suspend), but what other examples do we have where, silently, instructions don't get executed after a call?

Also, maybe this would become moot if we address https://discourse.llvm.org/t/coro-pre-split-handling-of-the-suspend-edge/75043 like @jyknight suggested (i.e. not even have the misleading edge)?

Reply:
> Also, maybe this would become moot if we address https://discourse.llvm.org/t/coro-pre-split-handling-of-the-suspend-edge/75043 like @jyknight suggested (i.e. not even have the misleading edge)?

But IIRC, it is still possible that we'll have code inserted between `llvm.coro.await.suspend.{.*}` and `llvm.coro.suspend`, which is the problem we're discussing.

> Sorry for insisting on this, it's maybe because I got "bitten" before (with the suspend), but what other examples do we have where, silently, instructions don't get executed after a call?

If I read correctly, @zmodem said he'd like to mention this in the doc or check it by assertions or verifiers. So it looks consensus to me?

Reply:
> Sorry for insisting on this

Not at all, I want us to resolve this, not paper over it. :)

>  it's maybe because I got "bitten" before (with the suspend)

But in that case the problem was kind of the opposite, right? Instructions between the `resume` and `suspend` blocked the musttail optimization. The problem was not whether those instructions would have been executed or not, but that they were inserted in a place where it wasn't allowed due to special constraints on this intrinsic.

The idea with my patch is to eliminate that problem by not having special restrictions about instructions after the intrinsic, they just won't be executed because control continues in the resumed function and doesn't come back.

> what other examples do we have where, silently, instructions don't get executed after a call?

This happens any time a call doesn't return.

> Also, maybe this would become moot if we address https://discourse.llvm.org/t/coro-pre-split-handling-of-the-suspend-edge/75043 like @jyknight suggested (i.e. not even have the misleading edge)?

Yes that's probably true, but that's a bigger change and I think my patch is a simpler solution to this specific problem: now that we have an intrinsic that corresponds directly to the symmetric transfer, let's just lower it as such directly.

> If I read correctly, @zmodem said he'd like to mention this in the doc or check it by assertions or verifiers. So it looks consensus to me?

No, I said that *if* we decide that instructions after `llvm.coro.await.suspend.handle` are not allowed, we should document and check it.

I still don't think we should do that, since it's a difficult invariant to maintain. We'd have to teach passes to respect it, keep updating a list of harmless instructions that are allowed (debug/lifetime intrinsics, ...) and we'd end up chasing down those assert/verifier failures forever.

With the current patch, we don't have to teach other passes anything. They already cannot assume that an unknown intrinsic call will return.

Reply:
> But in that case the problem was kind of the opposite, right? Instructions between the resume and suspend blocked the musttail optimization. The problem was not whether those instructions would have been executed or not, but that they were inserted in a place where it wasn't allowed due to special constraints on this intrinsic.
> The idea with my patch is to eliminate that problem by not having special restrictions about instructions after the intrinsic, they just won't be executed because control continues in the resumed function and doesn't come back.

I think I see where the disconnect comes from. You're referring to the desired outcome for coro: tailcalls being ensured. I agree with that, but I'm concerned with maintenance: suppose some other pass does what PGOInstrument did, in the future. Just like with coro, the potential for a undesired interaction between coro and their pass isn't known to them. How easy would it be to discover they made incorrect assumptions in their pass, diagnose and fix?

Reply:
I'm also concerned with maintenance, that's why I'm pushing back against adding special rules around this intrinsic.

Let's say that we did add an assert checking for non-trivial instructions after `llvm.coro.await.suspend.handle`. If someone were to write a pass which inserts instrumentation after each call instruction to see if the call returns, they now need to treat coros as a special case. (Assuming they notice, by testing their pass on a coroutine.)

It's better if other passes don't have to have special cases for coroutines. I think that's the case with my current patch, because it fits within the existing IR model: function calls don't necessarily return, and when they don't, instructions after the call won't run and the optimizer may remove them. Here's a concrete example: https://godbolt.org/z/M65Ebcfh5

Reply:
In your example, if they derive some result based on the initial set of instrumentations they insert, they may produce reports that say "incomplete coverage" (for example). Some level of knowledge ends up being passed down to someone - maybe the user, maybe the analyzer of the data.

But I get the idea that maybe the domain is OK with loss of instructions, so the assert could be annoying. Dunno, only saw the example where it wasn't.

How about an ORE? The output could technically be usable by a tool in a fictitious coverage case, it offers more focused observability than dumping IR before/after in a debugging case, and it serves as self-documentation in the pass itself that "ya, we're dropping stuff here. It's intentional". The ORE wouldn't need to be too complex, maybe if we detect non-debug (and maybe also exclude lifetime) instructions, ORE out. WDYT?

Reply:
An optimization remark would be less disruptive for sure, but I'm still not sure that it's right. Even though remarks are much lower level than regular diagnostics, they're still targeted at the compiler end user. If we emit a remark that says "deleted instructions after llvm.coro.await.suspend.handle call", what would the user do with that?

Reply:
For a fictitious example using some concrete actors: they could author a coverage analysis tool that would know to look for that and not misreport (say) 66% coverage because one of the edges was never traversed (because it never really existed)?

Reply:
That still sounds like the remark would not target the compiler user, but the developer of the compiler tool -- and that's not really what optimization remarks are for.

The more I think about this, the less I like the idea of a remark. A "deleted instructions after llvm.coro.await.suspend.handle call" remark would be pure noise for a normal optimization remarks user. Plus we would need to maintain the code which to emit it, including tests for when it should trigger or not.

I think my patch makes the semantic of llvm.coro.await.suspend.handle clear in the doc, I think that behavior is compatible with how calls in the IR already works, and I think existing passes will handle it well. I don't think it needs special handling.

Reply:
> That still sounds like the remark would not target the compiler user, but the developer of the compiler tool -- and that's not really what optimization remarks are for.

Not sure if that's the case anymore, they do have a yaml outputter.

But that's besides the point. I think I can see this your way: in our running example, if I worked on coverage, and knew to write the ORE post-processor, I would also know to fix my coverage instrumenter to recognize not to place instructions in the wrong place. (I realize there are other cases where passes are OK to stay agnostic despite losing instructions here, but in the coverage scenario that's not the case any way you look at it). The main thing is how to figure I may have a problem, and really the fix there is "integration testing", really.